walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,79 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        final long meetingDuration = request.getDuration(); // in minutes
+        Collection<String> requestAttendees = request.getAttendees();
+
+        Collection<TimeRange> availableTimes = new ArrayList<>();
+
+        if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+            return availableTimes;
+        }
+        if (events.size() == 0) {
+            availableTimes.add(TimeRange.WHOLE_DAY);
+            return availableTimes;
+        }
+
+        for (Event event : events) {
+            TimeRange when = event.getWhen();
+            Set<String> eventAttendees = event.getAttendees();
+
+            boolean hasNoCommonAttendees = Collections.disjoint(requestAttendees, eventAttendees);
+             
+            // if no event attendees need to go to requested meeting
+            if (hasNoCommonAttendees) {
+                if (availableTimes.size() == 0) {
+                    availableTimes.add(TimeRange.WHOLE_DAY);
+                }
+                continue;
+            }
+
+            // creates TimeRange available before and after event
+            TimeRange availableBefore = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, when.start(), false);
+            TimeRange availableAfter = TimeRange.fromStartEnd(when.end(), TimeRange.END_OF_DAY, true);
+            
+            availableTimes = returnOverlaps(availableBefore, availableAfter, availableTimes, meetingDuration);
+        }
+
+        return availableTimes;
+    }
+
+    private Collection<TimeRange> returnOverlaps(TimeRange before, TimeRange after, Collection<TimeRange> available, long duration) {
```

Comment:
nit: rename "available" to "availableTimes"

Reply:
also rename "duration" to meetingDuration so that it's clearer which object it refers to (e.g. Event vs. Meeting)

Reply:
and also beforeTimeRange and afterTimeRange